compiler/rustc_span/src/hygiene.rs

```diff
@@ -1474,6 +1493,28 @@ pub fn decode_syntax_context<D: Decoder, F: FnOnce(&mut D, u32) -> SyntaxContext
             // This is a completely new context.
             // Overwrite its placeholder data with our decoded data.
             None => {
+                let (opaque, opaque_and_semitransparent) =
+                    if ctxt_key.outer_transparency >= Transparency::Opaque {
+                        (pending_ctxt, pending_ctxt)
+                    } else if ctxt_key.outer_transparency >= Transparency::SemiTransparent {
+                        let opaque =
+                            hygiene_data.syntax_context_data[ctxt_key.parent.0 as usize].opaque;
+                        (opaque, pending_ctxt)
+                    } else {
+                        let opaque =
+                            hygiene_data.syntax_context_data[ctxt_key.parent.0 as usize].opaque;
+                        let opaque_and_semitransparent = hygiene_data.syntax_context_data
+                            [ctxt_key.parent.0 as usize]
+                            .opaque_and_semitransparent;
+                        (opaque, opaque_and_semitransparent)
+                    };
+                let ctxt_data = SyntaxContextData::new(
+                    ctxt_key,
+                    opaque,
+                    opaque_and_semitransparent,
+                    kw::DollarCrate,
+                );
```

Comment:
I don't think this reproduces the logic in `apply_mark_internal` correctly.

This code needs to reuse `apply_mark_internal` directly, except it will need to put the resulting data into `pending_ctxt` and not into the end of the `syntax_context_data` vector.
Something like this https://github.com/petrochenkov/rust/commit/0990edd67c0d1c5ea0b6572dcdbd5e11c8bfc5f3.

Reply:
(The `assert_eq` below will also need to be moved to `apply_mark_internal`.)

Reply:
> This code needs to reuse apply_mark_internal directly

I don't quite agree that we can directly use `apply_mark_internal` here, because the `pending_ctxt` needs to serve a certain purpose, and calling `apply_mark_internal` directly would negate that effect.

Later, I will try some approaches and provide a more convincing explanation (it might take some time as work has been very hectic lately).

Reply:
Actually, I think after this PR we should be able to get rid of the context id reservation (`pending_contest`) entirely - https://github.com/rust-lang/rust/pull/130324#event-14255590920, I just didn't want to include it into this PR because the logic here is complex and it's better to make and review changes in small steps.
Without context reserving `apply_mark_internal` will just always return a fresh `SyntaxContext`, both in `apply_mark` and in `decode_syntax_context`.

Reply:
> > This code needs to reuse apply_mark_internal directly
> 
> I don't quite agree that we can directly use `apply_mark_internal` here, because the `pending_ctxt` needs to serve a certain purpose, and calling `apply_mark_internal` directly would negate that effect.
> 
> Later, I will try some approaches and provide a more convincing explanation (it might take some time as work has been very hectic lately).

I suspect you mean some intermediate steps in `apply_mark_internal` should also reuse `pending_ctxt` if they create `SyntaxContextData` matching it, not just the final step. Yeah, then https://github.com/petrochenkov/rust/commit/0990edd67c0d1c5ea0b6572dcdbd5e11c8bfc5f3 is not enough.

Reply:
The version of `apply_mark_internal` from https://github.com/rust-lang/rust/pull/139083 should be easier to reuse for decoding.